core/embed/trezorhal/stm32f4/touch/ft6x36.c

```diff
@@ -74,227 +148,278 @@ static void touch_default_pin_state(void) {
-                            TOUCH_PACKET_SIZE, 1)) {
-    return 0;  // read failure
-  }
+  // Extract number of touches (0, 1, 2) or 0x0F before
+  // the first touch (tested with FT6206)
+  uint8_t nb_touches = regs[FT6X63_REG_TD_STATUS] & 0x0F;
 
-  last_event_time = hal_ticks_ms();
+  // Extract event flags (one of press down, contact, lift up)
+  uint8_t flags = regs[FT6X63_REG_P1_XH] & FT6X63_EVENT_MASK;
 
-  if (0 == memcmp(previous_touch_data, touch_data, TOUCH_PACKET_SIZE)) {
-    return 0;  // same data, filter it out
-  } else {
-    memcpy(previous_touch_data, touch_data, TOUCH_PACKET_SIZE);
-  }
+  // Extract touch coordinates
+  uint16_t x = ((regs[FT6X63_REG_P1_XH] & 0x0F) << 8) | regs[FT6X63_REG_P1_XL];
+  uint16_t y = ((regs[FT6X63_REG_P1_YH] & 0x0F) << 8) | regs[FT6X63_REG_P1_YL];
+
+  uint32_t event = 0;
 
-  const uint32_t number_of_touch_points =
-      touch_data[2] & 0x0F;  // valid values are 0, 1, 2 (invalid 0xF before
-                             // first touch) (tested with FT6206)
-  const uint32_t event_flag = touch_data[3] & 0xC0;
-  if (touch_data[1] == GESTURE_NO_GESTURE) {
-    xy = touch_pack_xy((X_POS_MSB << 8) | X_POS_LSB,
-                       (Y_POS_MSB << 8) | Y_POS_LSB);
-    if ((number_of_touch_points == 1) && (event_flag == EVENT_PRESS_DOWN)) {
-      touching = 1;
-      return TOUCH_START | xy;
-    } else if ((number_of_touch_points == 1) && (event_flag == EVENT_CONTACT)) {
-      if (touching) {
-        return TOUCH_MOVE | xy;
-      } else {
-        touching = 1;
-        return TOUCH_START | xy;
+  uint32_t xy = touch_pack_xy(x, y);
+
+  if ((nb_touches == 1) && (flags == FT6X63_EVENT_PRESS_DOWN)) {
+    if (!driver->pressed) {
+      // Finger was just pressed down
+      event = TOUCH_START | xy;
+    } else {
+      // It looks like we have missed the lift up event
+      // We should send the TOUCH_END event here with old coordinates
+      event = TOUCH_END | touch_pack_xy(driver->last_x, driver->last_y);
+    }
+  } else if ((nb_touches == 1) && (flags == FT6X63_EVENT_CONTACT)) {
+    if (driver->pressed) {
+      if ((x != driver->last_x) || (y != driver->last_y)) {
+        // Report the move event only if the coordinates
+        // have changed
+        event = TOUCH_MOVE | xy;
+      }
+    } else {
+      // We have missed the press down event, we have to simulate it.
+      // But ensure we don't simulate TOUCH_START if touch_get_event() is not
+      // called frequently enough to not produce false events.
+      if (!starving) {
+        event = TOUCH_START | xy;
       }
-    } else if ((number_of_touch_points == 0) && (event_flag == EVENT_LIFT_UP)) {
-      touching = 0;
-      return TOUCH_END | xy;
+    }
+  } else if ((nb_touches == 0) && (flags == FT6X63_EVENT_LIFT_UP)) {
+    if (driver->pressed) {
+      // Finger was just lifted up
+      event = TOUCH_END | xy;
+    } else {
+      // Finger is still lifted up. Since we have already sent the
```

Comment:
I also considered this and have to admit I am not absolutely sure. On one side, this branch blocks sending `TOUCH_END` twice (it should not happen because of the interrupt line check, but who knows...). On the other side, it is really missed the `PRESS_DOWN` event. But if it is, then the touch duration was really short—just 10ms (with no `CONTACT` event in between).

I would rather keep it as it is since the proper solution requires simulating `TOUCH_START` and subsequent `TOUCH_END` in the next call, which is more complicated and probably brings no real value. We can of course discuss it further.

Reply:
I am fine with that, just please adjust the comment so its clear that its intentional

Reply:
commented, see a96ade8a4b2b371145185d554c3ce3ff0feb8299